pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java

```diff
@@ -97,6 +117,124 @@ private String getPhoneNumberFromActivities(List<Activities> activities) {
+        Optional<FieldSettingsDto> refSource = fieldSettingsDao.getFieldSettingsByFieldTypeAndColumnName(
+                "RGP_MEDICAL_RECORDS_GROUP", "REF_SOURCE");
+
+        // for REF_SOURCE, the details column hold the mapping between DSS referral sources (answers to FIND_OUT
+        // enrollment question) to DSM ref sources
+        Optional<String> details = refSource.map(FieldSettingsDto::getDetails);
+        if (details.isEmpty()) {
+            throw new RuntimeException("FieldSettings 'details' is empty for RGP_MEDICAL_RECORDS_GROUP REF_SOURCE");
+        }
+        Optional<String> possibleValues = refSource.map(FieldSettingsDto::getPossibleValues);
+        if (possibleValues.isEmpty()) {
+            throw new RuntimeException("FieldSettings 'possibleValues' is empty for " + "" +
+                    "RGP_MEDICAL_RECORDS_GROUP REF_SOURCE");
+        }
+
+        return this.deriveReferralSource(sources, details.get(), possibleValues.get());
+     }
+
+    /**
+     * Given the referral sources provided using DSS FOUND_OUT IDs, get the corresponding DSM referral source ID
+     *
+     * @param sources referral sources provided as DSS stable IDs
+     * @param refDetails JSON string of FieldSettings REF_SOURCE details column
+     * @param refPossibleValues JSON string of FieldSettings REF_SOURCE possibleValues column
+     * @return a referral source ID
+     * @throws RuntimeException If the expected REF_SOURCE values are not present or the referral source mapping is
+     * out of sync, the method logs errors, but will throw if the provided referral sources cannot be mapped.
+     */
+    protected String deriveReferralSource(List<String> sources, String refDetails, String refPossibleValues) {
+        // algorithm: if more than one answer chosen, then use REF_SOURCE MORE_THAN_ONE,
+        // if no answer, or if the question is missing from enrollment, etc., then use REF_SOURCE NA
+
+        // details column holds a map of FIND_OUT answers to REF_SOURCE IDs
+        Map<String, String> refMap = ObjectMapperSingleton.readValue(
+                refDetails, new TypeReference<Map<String, String>>() {});
+
+        // get the REF_SOURCE IDs to verify the map is still in sync
+        List<Map<String, String>> refValues = ObjectMapperSingleton.readValue(
+                refPossibleValues, new TypeReference<List<Map<String, String>>>() {});
+
+        Set<String> refIDs = refValues.stream().map(m -> m.get("value")).collect(Collectors.toSet());
+        Set<String> refMapValues = new HashSet<>(refMap.values());
+        // "NA" and "MORE_THAN_ONE" are not mapped
+        if (refMapValues.size()+2 != refIDs.size() || !refIDs.containsAll(refMapValues)) {
+            // if IDs have diverged, don't stop this operation: log error and see if something is salvageable (below)
+            logger.error("RGP_MEDICAL_RECORDS_GROUP REF_SOURCE 'possibleValues' do not match REF_SOURCE map in " +
```

Comment:
Log level appropriate?  Should it be WARN?
The question I'm thinking about in my head for these types of things: "Should we get an entry in the slack channel for this or is it somewhat expected and should we just log it?"

Reply:
I'll explain this better in the comment, but it is definitely something we should get alerted on. It's a runtime internal error, but there's not much recourse at this point (pubsub message handler) so I think it makes sense to press forward if we are able.